File: epi_judge_python/is_valid_parenthesization.py
from test_framework import generic_test


def is_well_formed(s: str) -> bool:
    stack = []
    matching = {'}' : '{', ']' : '[', ')' : '('}
    for x in s:
        if x in matching.keys():
            if len(stack) > 0 and stack[-1] == matching[x]:
                stack.pop()
            else:
                return False
        else:
            stack.append(x)
    return True if len(stack) == 0 else False


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('is_valid_parenthesization.py',
                                       'is_valid_parenthesization.tsv',
                                       is_well_formed))

File: epi_judge_cpp/is_valid_parenthesization.cc
#include <string>

#include "test_framework/generic_test.h"
using std::string;
bool IsWellFormed(const string &s) {
  std::stack<char> parens;
  for (auto &ch : s) {
    if (ch == '(' || ch == '{' || ch == '[') {
      parens.push(ch);
    } else {
      if (parens.empty()) {
        return false;
      }
      if ((ch == ')' && parens.top() != '(') ||
          (ch == '}' && parens.top() != '{') ||
          (ch == ']' && parens.top() != '[')) {
        return false;
      }
      parens.pop();
    }
  }
  return parens.empty();
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"s"};
  return GenericTestMain(args, "is_valid_parenthesization.cc",
                         "is_valid_parenthesization.tsv", &IsWellFormed,
                         DefaultComparator{}, param_names);
}
